README.md

```diff
@@ -1,12 +1,332 @@
+and determines the most efficient routine to use (the Woodbury formula).
+This way, we can get a efficient ($\mathcal O(N)$) solve while abstracting away all of the details.
+
+Crucially, $\boldsymbol A$ is never explicitly instantiated as a matrix, which makes it possible to scale
+to very large operators without running out of memory:
+
+```python
+C = torch.randn(10000000, 20)
+d = torch.randn(10000000)
+b = torch.randn(10000000)
+A = RootLinearOperator(C) + DiagLinearOperator(d)  # represents a 10M x 10M matrix!
+torch.linalg.solve(A, d, b)  # computes A^{-1} b efficiently!
 ```
+
+### What is a Linear Operator?
+A linear operator is a generalization of a matrix.
+It is a linear function that is defined in by its application to a vector.
+The most common linear operators are (potentially structured) matrices,
+where the function applying them to a vector are (potentially efficient)
+matrix-vector multiplication routines.
+
+In code, a `LinearOperator` is a class that
+1. specifies the tensor(s) needed to define the LinearOperator,
+1. specifies a `_matmul` function (how the LinearOperator is applied to a vector),
+1. specifies a `_size` function (how big is the LinearOperator if it is represented as a matrix, or batch of matrices), and
+1. specifies a `_transpose_nonbatch` function (the adjoint of the LinearOperator).
+
+For example:
+
+```python
+class DiagLinearOperator(linear_operator.LinearOperator):
+    r"""
+    A LinearOperator representing a diagonal matrix.
+    """
+    def __init__(self, diag):
+        # diag: the vector that defines the diagonal of the matrix
+        self.diag = diag
+
+    def _matmul(self):
+        return torch.Size([*self.diag.shape, self.diag.size(-1)])
+
+    def _size(self):
+        return torch.Size([*self.diag.shape, self.diag.size(-1)])
+
+    def _transpose_nonbatch(self):
+        return self  # Diagonal matrices are symmetric
+
+
+# ...
+
+D = DiagLinearOperator(torch.tensor([1., 2., 3.])
+# Represents the matrix
+#   [[1., 0., 0.],
+#    [0., 2., 0.],
+#    [0., 0., 3.]]
+torch.matmul(D, torch.tensor([4., 5., 6.])
+# Returns [4., 10., 18.]
+```
+
+Other functions (e.g. `logdet`, `eigh`, etc.) can be defined as needed if efficient routines exist.
+
+#### Why is This Useful?
+While `_matmul`, `_size`, and `_transpose_nonbatch` might seem like a limited set of functions,
+it turns out that most functions on the `torch` and `torch.linalg` namespaces can be efficiently implemented
+using only these three primitative functions.
+
+Moreover, because `_matmul` is a linear function, it is very easy to compose linear operators in various ways.
+For example: adding two linear operators (`SumLinearOperator`) just requires adding the output of their `_matmul` functions.
+This makes it possible to define very complex compositional structures that still yield efficient linear algebraic routines.
```

Comment:
We should also talk about the fact that the product (not in the mathematical sense) of an operation on one or more `LinearOperator` objects can again return a `LinearOperator`, i.e. that the operations are structure preserving. E.g. we can give the example of summing two `DiagLinearOperator`s.

Reply:
This exists in the [Using LinearOperator objects](https://github.com/cornellius-gp/linear_operator/blob/readme/README.md#using-linearoperator-objects) section - but do you have any suggestions for where to put it more prominently?

Reply:
I think it's fine to just add a comment here in this list that states this (possibly referring the reader to the details below)